manual/manual/cmds/instrumented-runtime.etex

```diff
@@ -0,0 +1,332 @@
+down to the time spent in specific parts of the garbage collection algorithm.
+
+Insight is also given on memory allocation and motion, recording
+the size of allocated memory blocks, as well as value promotions from the
+{\em minor heap} to the {\em major heap}.
+
+\section{s:instr-runtime-overview}{Overview}
+
+Once compiled and linked with the instrumented runtime, any OCaml program
+can generate {\em trace files} that can then be read
+and analyzed by users in order to understand specific runtime behaviors.
+
+The {\em Common Trace Format} is the general purpose binary
+format used to encode traces.
+A complete trace is comprised of:
+\begin{itemize}
+\item a {\em metadata file}, part of the OCaml distribution
+\item and a {\em trace file}, generated by the runtime\
+  in the program being traced.
+\end{itemize}
+
+For more information on the {\em Common Trace Format}, see
+\ifouthtml
+\ahref{https://diamon.org/ctf/}{https://diamon.org/ctf/}.
+\else
+{\tt https://diamon.org/ctf/}
+\fi
+
+\section{s:instr-runtime-enabling}{Enabling runtime instrumentation}
+
+
+For the following examples, we will use the following example program:
+
+\begin{verbatim}
+       module SMap = Map.Make(String)
+
+       let s i = String.make 512 (Char.chr (i mod 256))
+
+       let clear map = SMap.fold (fun k _ m -> SMap.remove k m) map map
+
+       let rec seq i =
+         if i = 0 then Seq.empty else fun () -> (Seq.Cons (i, seq (i - 1)))
+
+       let () =
+         seq 1_000_000
+         |> Seq.fold_left (fun m i -> SMap.add (s i) i m) SMap.empty
+         |> clear
+         |> ignore
+
+\end{verbatim}
+
+The next step is to compile and link the program with the instrumented runtime.
+This can be done by using the {\tt -runtime-variant} flag:
+
+\begin{verbatim}
+       ocamlopt -runtime-variant i program.ml -o program
+\end{verbatim}
+
+Note that the instrumented runtime is an alternative runtime for OCaml
+programs. It is only referenced during the linking stage of the final
+executable. This means that the compilation stage does not need to be altered
+to enable instrumentation.
+
+The resulting program can then be traced by running it with the environment
+variable {\tt OCAML\_EVENTLOG\_ENABLED}:
+
+\begin{verbatim}
+        OCAML_EVENTLOG_ENABLED=1 ./program
+\end{verbatim}
+
+During execution, a trace file will be generated in the
+program's current working directory.
+
+\subsubsection*{sss:instr-runtime-build-more}{More build examples}
+
+When using the {\em dune} build system, this compiler invocation can be
+replicated using the {\tt flags} {\tt stanza} when building an executable.
+
+\begin{verbatim}
+       (executable
+         (name program)
+         (flags "-runtime-variant=i"))
+\end{verbatim}
+
+The instrumented runtime can also be used with the OCaml bytecode interpreter.
+This can be done by either compiling the program using the
+{\tt -runtime-variant=i} flag when linking the program {\tt ocamlc}, or by running the generated
```

Comment:
Another point in 
> by either compiling the program using the {\tt -runtime-variant=i} flag when linking the program {\tt ocamlc}

the phrase  "compiling the program  ...when linking the program" seems confusing when you carefully distinguished the compiling and linking phase above.

Reply:
I think this sentence can be shorted to
> This can be done by either using the "-runtime-variant=i" flag when linking the program with {\tt ocamlc},

Reply:
The shortened sentence sounds better indeed.